pex/cli/commands/lock.py

```diff
@@ -3,28 +3,45 @@
 
 from __future__ import absolute_import
 
-from argparse import ArgumentParser, _ActionsContainer, _SubParsersAction
+from argparse import ArgumentParser, _ActionsContainer
+from collections import defaultdict
 
+import pex.cli.commands.lockfile
 from pex import resolver
 from pex.cli.command import BuildTimeCommand
-from pex.commands.command import Ok, OutputMixin, Result
+from pex.cli.commands import lockfile
+from pex.cli.commands.lockfile import Lockfile, json_codec
+from pex.commands.command import Error, JsonMixin, Ok, OutputMixin, Result
+from pex.common import pluralize
+from pex.distribution_target import DistributionTarget
+from pex.enum import Enum
+from pex.requirements import LocalProjectRequirement
 from pex.resolve import requirement_options, resolver_options, target_options
-from pex.resolve.locked_resolve import LockConfiguration, LockStyle
+from pex.resolve.locked_resolve import LockConfiguration, LockedResolve, LockStyle
+from pex.third_party.pkg_resources import Requirement
+from pex.tracer import TRACER
+from pex.typing import TYPE_CHECKING
 from pex.variables import ENV
+from pex.version import __version__
 
+if TYPE_CHECKING:
+    from typing import List, DefaultDict
 
-class Lock(OutputMixin, BuildTimeCommand):
+
+class LockFormat(Enum["LockFormat.Value"]):
+    class Value(Enum.Value):
+        pass
+
+    PIP = Value("pip")
+    PEP_665 = Value("pep-665")
```

Comment:
So, is the implication of `PEP-665` being an export format that it didn't work as an internal representation of the file? That would likely be useful feedback for them.

Reply:
It is not, I was just making space for someone else to pick up #1403 in a clear way. The spec isn't real though yet, so at bottom, it makes most sense for Pex to store its lock with all the data it needs right now in a real format. The `pex_version` field in that format allows us to fully switch to PEP-665 as a native format if it makes sense in the future. That aside, the code doesn't currently produce enough data to produce a valid PEP 665 lock yet. That will need to combine the metadata from the downloaded local distributions in a Downloaded result with the LockedResolves found in a Downloaded to produce the needs / needed_by for each project.

Reply:
To put a finer point on this: way back when PEP 665 was brought up I reviewed it and the thread at that time. My conclusion then was that it covered everything Pex needed. That conclusion was based on my lock experiments - so real data - and has not changed.

Reply:
> It is not, I was just making space for someone else to pick up #1403 in a clear way. The spec isn't real though yet, so at bottom, it makes most sense for Pex to store its lock with all the data it needs right now in a real format. The `pex_version` field in that format allows us to fully switch to PEP-665 as a native format if it makes sense in the future.

Chickens and eggs. =)

> That aside, the code doesn't currently produce enough data to produce a valid PEP 665 lock yet. That will need to combine the metadata from the downloaded local distributions in a Downloaded result with the LockedResolves found in a Downloaded to produce the needs / needed_by for each project.

Got it. Thanks!

Reply:
> Chickens and eggs. =)

I'm not ruling out being God here and upending chickens and eggs, its just that it seems most appropriate to do this via an _export_ to PEP-665 being the 1st instance of it in the wild and not the fundamental underlying lock format for Pex. Hopefully that makes sense and you agree its the only sane way to proceed with vaporware.

Reply:
Yea, not a blocking comment. Talked with @chrisjrn about this at some point, but even if PEP-665 had had a reference implementation of the format, it's unlikely that that reference implementation would have been drop-in for any consumers (PEX included), so the vaporware-ness seems like an optimistic guess that someone will try implementing the spec and report back about whether it was possible/useful.

Reply:
For a bit of clarity I'll rename this to `ExportFormat`.  The export does not even need to be to a proper lock, it could be to a pinned constraints file for example.